dhis-2/dhis-support/dhis-support-db-migration/src/main/resources/org/hisp/dhis/db/migration/2.42/V2_42_26__Add_not_null_constraint_to_trackedentity.sql

```diff
@@ -0,0 +1,39 @@
+DO $$
+DECLARE
+results RECORD;
+inconsistent_records_count INTEGER;
+BEGIN
+SELECT COUNT(*) INTO inconsistent_records_count FROM trackedentity WHERE trackedentitytypeid IS NULL;
+
+IF inconsistent_records_count > 0 THEN
+    RAISE NOTICE 'Inconsistencies found: %', inconsistent_records_count;
+
+    FOR results IN
+    SELECT programid, trackedentityid FROM enrollment WHERE trackedentityid IN (
+        SELECT trackedentityid
+        FROM trackedentity
+        WHERE trackedentitytypeid IS NULL
+    )
+    LOOP
+    UPDATE trackedentity
```

Comment:
Here we shouldn't update any record.
We should only apply the constraint.
The we can suggest such a script in the migration notes

Reply:
there is a lower possibility of errors if we handle the task ourselves rather than leaving it to the user. Therefore, we should aim to complete as much of the migration process as possible and leave as little as necessary for the user to manage.

Reply:
For this kind of process that deal with data we need to be much more clear with the user and very very careful on any step.
Here you are fixing inconsistent data in a correct way, getting the tracked entity type from any underling enrollment but this is not the only issue that could happen and this SQL is running in a transaction (even without specifically start the transaction as done here, flyway will wrap this in a transaction), so if there is any inconsistent data that cannot be fixed in the way you provided, no data is fixed.
Than the user is sent to the guide and there we are not providing any explanation of what we attempted in the migration and we are not providing any option of fixing the data, only a script to delete tracked entitities and with all the enrollments and events.

To resume:
If the user has a number x of "fixable" tracked entities and one that is not automatically fixable, we are deleting ALL the tracked entities with inconsistent data.